### ISSUE TITLE
Reactive variables SSR concurrency

### ISSUE THREAD

**Issue by vxcamiloxv** [apollographql/apollo-client#8924]
**Intended outcome:**
Set a reactive variable in SSR an send it to client-side, If in the render of a component we set a reactive variable they should keep the value only until the cache is clean or Apollo instance is begin destroyed on SSR or CSR

**Actual outcome:**
The reactive variable keep the same value across request on SSR event if the apollo client even if stopped / destroyed or the cache was cleaned/reset, it becomes a global variable in server, even if you don't need to send it to the client the value becomes a global variable 

**How to reproduce the issue:**
I created a repo with Next.js https://github.com/vxcamiloxv/reactive-variables-ssr-concurrency-issue to reproduce the issue, as you can see there is a reactive variable and I set a value on SSR and then sent it to client side.

**Versions**
Tested with 3.4.16 and 3.4.10

**Comment by andrefagl** [comment]
@vxcamiloxv could you please provide your apollo configuration? It is not present in the reproduction repo :(

**Comment by vxcamiloxv** [comment]
Sorry, I thought was uploaded it, this is the configuration https://github.com/vxcamiloxv/reactive-variables-ssr-concurrency-issue/blob/main/lib/apollo.ts, it doesn't have anything special, happens with any configuration

**Comment by joaopedrodcf** [comment]
Did someone achieve to fix this? The Reactive variables are set on the server and persist through sessions even if in each call to SSR we create a new instance of apollo client just like the example above ☝️ https://github.com/apollographql/apollo-client/issues/8924#issuecomment-962283770

Is there a way to reset all reactive variables in the SSR?

**Comment by andrefagl** [comment]
Was digging a bit through the internals and found this test: https://github.com/apollographql/apollo-client/blob/015c40314429b0a7f5082a25da24d3dc208df383/src/cache/inmemory/__tests__/cache.ts#L3506 where an RV is created and then call `rv.forgetCache(cache)`.

It seems that `forgetCache` only prevents the RV update to trigger a broadcast and does not actually clear its value.

It would be nice to have the input of a maintainer if there is a way to delete RVs between sessions as it should be expected to happen when creating a new ApolloClient instance.

**Comment by ianvieira** [comment]
After some digging, I think I was able to find the issue. Although all the sophisticated cache layers the RV has, they are more relevant to broadcast changes than to the value itself. 

Looking at the code of `makeVar`, the value itself doesn't come from a cache or anything else. It is, in fact, the `value` from the first `makeVar`, and the updates are just updating this particular variable and use the cache structure to broadcast the update, as in the code below:

https://github.com/apollographql/apollo-client/blob/cd6f4061693d66091384b43f57180dc88529c00b/src/cache/inmemory/reactiveVars.ts#L51-L84

This code is fine as long as you run the `makeVar` on all the renders. Which we technically do, if it wasn't for Node itself.

This is how our reactive variables file looks like:

```ts
import { makeVar } from '@apollo/client';

export const variableA = makeVar<string>('');
export const variableB = makeVar<string | null>(null);
```

And then we use `variableA` and `variableB` through imports on our code. The problem is that [Node caches each module required](https://bambielli.com/til/2017-04-30-node-require-cache/), for performance improvements, this means the code running on the module to generate the export is cached too.

Between renderings, we have the same import result, this makes the `value` in `makeVar` persist, and when the next rendering is going to use the reactive variable, it will already have an assigned value. 

We don't use Next, but I believe we both use the same strategy, we bundle an SSR version of the code, which exports a function and we run this function. I did a little experiment on our code and I was able to mitigate this issue by doing:

```js
// Eval require makes it possible to use absolute paths
const globalRequire = eval('require');
// Removes the file from node cache 
delete globalRequire.cache[serverFilePath];
// Require the file without cache
const renderer = globalRequire(serverFilePath).default;
```

As this forces the entire module to run again, it will avoid the cache and generate fresh values. This workaround cannot be used in production, because in our case it would add 180ms in a 60ms operation, which is totally off the table. 

I believe a permanent solution would be to save reactive variables on the cache itself, this way the value can be destroyed along with the `ApolloCache` instance which we guarantee is a new one on each rendering, something we cannot do with Reactive Variables. This also can solve other issues as the RV is not being extracted, as technically speaking, their value is really not coming from the cache, the cache is just "notified" that it has a new value (from what I can understand).

We are available for helping to solve this issue, as it represents a major problem for us, be it by helping with the fix or with a minimum reproducible repo (which is challenging for us, as we are part of a closed source project).

**Comment by serle** [comment]
I agree with @ianvieira. I think that there should be a concept of "locally/client" scoped reactive variables created through `client.makeVar()`. In ssr mode the reactive var's should not be reactive at all as reactivity does not make sense server-side, but they would need so that the same components can render as is both on the server and client side. Reactivity could be controlled based on the ssr mode of the "parent" client. The same comment about reactivity based on the ssr mode of the client should apply to the other hooks like useQuery etc.

**Comment by nersoh** [comment]
Any update on this?

**Comment by phryneas** [comment]
I think, this issue is mostly based on a misunderstanding: Reactive Variables can be used without a cache, just as reactive variables. They carry their own value. So you have to handle them like you would handle every other variable in a SSR environment.

You can always just call
```js
const myVar = makeVar()
myVar(5)
console.log(myVar())
```
completely without involving an Apollo Client or Apollo Cache - and many people use them like that just to share values between React components with `useReactiveVar`.

Now, in a long-running node process that serves multiple requests (often in parallel) like a NextJS SSR environment, every variable you create in module scope (and possibly `export` and `import`) will be shared between all requests. That's why we recommend not to create a global `ApolloClient` in those environments, but one per request.  
That same advice goes for every other global value that you do not want to share between requests - also for reactive variables.  
If you want to have those scoped "by React render" (and you probably should), you cannot create those globally, but you have to create them in your `App` component and put them into something like React context and use that as a dependency injection layer so you have separate reactive variables per request (or "Application rendered in parallel").

This is not specific to Apollo Client or Reactive Vars.  
You will have to do similar things with ever other state management approach - this is due to the way NextJS and other SSR solutions work!

Bottom line: I don't think there is anything actionable we as the Apollo Client team can do here.

**Comment by ianvieira** [comment]
@phryneas I don't completely agree this is anything that is actionable on Apollo Client, especially with React 18 changes and how hydration works in general.

I agree Reactive Variables can be used as a standalone feature, apart from the ApolloClient itself, but at the same in this use-case it's basically a Observer, for which we could use a library only for that (as Rx.js or even Zen Observable, that Apollo uses). Apollo can keep this use-case support as it is but I don't believe this is the primary use-case in most of the time.

In my perspective as an Apollo Client user, the primary use-case for Reactive Variables is changing query parameters with automatic query invalidation, which directly connects with the need of an Apollo Client. I agree it's feasible to do all this boilerplate as you're suggesting but I also believe Apollo Client could provide another API merging both concepts, where we can create Reactive Variables which their values are stored on the cache instance. This would make much simpler the invalidation between requests (as we do destroy the cache and the client between requests) and also the extraction, making simpler the hydration, especially on React 18 scenario where this is streamed to the client and things get a little bit more tricky to handle.

All this work you described of creating a provider, storing the variables, propagating them, etc is basically what Apollo already does with their own provider and own store so I believe would be beneficial if we could reuse the same structure for a feature provided by Apollo Client itself, especially because this boilerplate gets more complex with React further iterations for a mix between server/client.

This also goes toward @serle suggestion and I believe would make much more integrated using Reactive Variables on the Apollo Client context instead of being a "separated data structure".

**Comment by vxcamiloxv** [comment]
@phryneas in the example that I provided the apollo client is created on each request in SSR https://github.com/vxcamiloxv/reactive-variables-ssr-concurrency-issue/blob/main/pages/_app.tsx#L46, it's not a global apollo client, additional for for testing purposes is also created on each render, https://github.com/vxcamiloxv/reactive-variables-ssr-concurrency-issue/blob/main/pages/_app.tsx#L17

**Comment by phryneas** [comment]
> especially with React 18 changes and how hydration works in general.

@ianvieira  could you explain what you mean by that? 
As I see it, in React 18 streaming SSR, you only have one run of data on the server that is hydrated on the client and thrown away on the server. You would not have anything "observable" that exists on the server, since you cannot even trigger a rerender on the server. You can choose to "rerender" a server component, but that means starting a completely new render with new data from zero - there is nothing long-running that exists between the first or second render, so observables on the server side don't seem to make any sense to me.

If we were to create a new paradigm for client-bound reactive variables as you suggest, those could not be called in any non-Apollo code any more without referencing an Apollo Client. So
```js
const currentValue = myVar()
myVar(newValue)
```
would have to become
```js
const currentValue = myVar(client)
myVar(client, newValue)
```
which would be very inconvenient and probably a show-stopper in many use cases.

With todays variables, you could write a helper function like
```js
const variables = new Symbol()
function getVariableByName(client, name, defaultValue) {
  const vars = client[variables] ??= {}
  return (vars[name] ??= makeVar(defaultValue))
}
function useVariableByName(name, defaultValue){
  return getVariableByName(useApolloClient(), name, defaultValue)
}
// usage in component:
const nameVar = useVariableByName(name)
```
to achieve something like store-bound "global" (in this case, indexed by name) variables in userland code - without any changes on our side.



@vxcamiloxv my point is that while you are creating one Apollo Client per Request, you are not creating a new reactive variable per request - and in any long-running node process that handles multiple requests, **anything global has to be created per request**. Your code is sharing global variables between multiple requests, and that's just a paradigm that you are not allowed to follow in NextJS.  
Again, this is not specific to Apollo Client in any way, this applies to every single `let`, `const`, or `var` at the top level in any of your files. They are shared between all requests.

That means, that you have to call `makeVar` inside the request context, not globally.

**Comment by vxcamiloxv** [comment]
@phryneas well, the expected is that due reactive variables are part of apollo client it should be work together, so if I destroy/stop apollo client form my understand the expected is clean any reference and the variables as welll, because it feels like another library apart of apollo in the sense that soy said , also the documentation don't said anything about that, about "this is just client-side" or "in server-side you should do a wrapper/helper", the code is following this https://www.apollographql.com/docs/react/local-state/reactive-variables, similar to the example there https://github.com/apollographql/ac3-state-management-examples/blob/master/apollo-local-state/src/cache.tsx.

So I think worth maybe that add that example of clarification in the docs

**Comment by phryneas** [comment]
@vxcamiloxv generally, I think the documentation is very clear that reactive variables are independent on the cache - the first two sentences of that link you gave reads:

> New in Apollo Client 3, reactive variables are a useful mechanism for representing local state **_outside of the Apollo Client cache_**. Because **_they're separate from the cache_**, reactive variables can store data of any type and structure, and you can interact with them anywhere in your application without using GraphQL syntax.

Nonetheless, I get that this can be a confusing thing, and unfortunately NextJS doesn't really do a good job of teaching the principle of "if you do SSR, don't have global variables", although it is very much a core problem there.

I am currently working on an extraneous package to add better support for RSC and SSR with NextJS. I will take this issue into consideration and we will try to add extra documentation for this issue when we release that.

As for the "wrapper" on server side, I have spent further thought about this, and I fear we will have to be more strict with our advice, and a wrapper might not be good either. Here's my thought process, I would be happy for feedback:
* As they are now, reactive vars on the server and reactive vars on the client are independent of each other. After Rehydrating data on the client, they will be back to their "initial value" on the client.
* We could add a mechanism to transport (at least serializable) values from the server to the client. This would need some api changes (`makeVar` would need an additional "identifier" argument), but it would theoretically possible.
* However, this would be purely unidirectional. There would not be a way of having changes of client-side reactive variables (or server-side changes from prior renders in the same "user session") reflected on the server (e.g. if you navigate, or if you trigger a RSC rerender).
* As a result, most use of reactive variables on the server seems unsafe to me.

At this point, I am considering that our advice should probably be on the lines of "avoid reactive variables on the server, but if you use them, only use their default value and never modify them on the server".  
I get that this is a restriction, but this is essentially just recommunicating restrictions that are generally in place when using RSC/SSR - again, none of this is specific to Apollo Client.

What are your thought on this?

**Comment by vxcamiloxv** [comment]
@phryneas For me is weird because Apollo client is used I think is mostly cases in SSR, and have a lot of references about ssr in the documentation and the expected if you are using a library like apollo is that all the provided helper/utility work together but I said reactive variables feels as you describe it like another library inside of the library, for me is like "make a helper to use our helper", but yeah  "avoid reactive variables on the server side render" would be useful

**Comment by phryneas** [comment]
@vxcamiloxv generally, Apollo Client is most useful "live" in the browser, where you have a long-running session and can profit most from a normalized cache.  
In any SSR environment, an Apollo Client instance will only exist for a very short amount of time, so it is mostly useful to use it there to gather some initial data that can later be used as a starting point of your in-browser cache. 

That is a use case we support, but going further from that, there are some limitations in place that are just there due to the nature of SSR being unidirectional. 
Data can only ever flow from the server to the client, so we wouldn't assume that people would use Reactive Variables in SSR scenarios, as the client value would very soon be different from the server value, and future SSR renders would necessarily be based on incorrect values.

Again, this is not really something we could support, even if we put every possible effort into it: It's how React SSR works. There is no shared session between Client and Server.

I'll discuss this further internally and we'll see that we put some warnings in place.

**Comment by ianvieira** [comment]
@phryneas With the recent announcements from Next.js, React and Apollo, I'd like to expand on my previous response. Apollo released an experimental version with Server Components support (https://www.apollographql.com/blog/announcement/frontend/using-apollo-client-with-next-js-13-releasing-an-official-library-to-support-the-app-router/).

This goes directly towards the answer I gave before, on the client itself was created an util `registerApolloClient `, which in Apollo words will:
> `registerApolloClient` will return a `getClient` function that allows you to safely get an instance of the Apollo Client that’s scoped to the current request, so you won’t end up sharing the Apollo Client cache within multiple requests.

This is an amazing step forward on React's direction but, again, we face the same problem with Reactive Variables, which currently are an integrant part of Apollo Client implementation and as I said before the best method to invalidate queries when we have changes. Although I understand the vision that this is separated from the client, the reality is that this is distributed with the client and is expected by the consumer on having the same level of support.

This issue will make impossible to adopt Server Components with React variables out-of-the-box, as we'd need to recreate most of the bootstrap Apollo already does to keep state between server and client in sync, that's why I suggested creating a new API which records the RVs on the cache and take leverage of tooling already created around this store, it gives us the simplicity of RVs APIs with the integration features we have on cache, creating the best development experience as possible.

This issue already made very challenging integrating the server-side with the client-side, now that React's direction (with Next support) is to split operations between server + client, this problem becomes almost impossible to overcome without Apollo's support and review on this API.

**Comment by phryneas** [comment]
Unfortunately, all of that makes it just even more complicated.

You have three environments now:

* React Server components. Here we have a `getClient` method that gives you the client for the current request.
* Client Components will also render on the server, but only on initial page load in your browser. Here you don't have any way of scoping something global "to the request", so the Apollo Client can only be passed down using Context. Trying something like `registerApolloClient` in this environment here would mix Apollo Clients between unrelated requests. `registerApolloClient` is a React Server Component api.
* The browser. Here you could have a a global variable with the Apollo Client, but since this is the same code that also runs during Client Component SSR, realistically you are also limited to React Context as well.

So, in 2 out of those 3 environments, there is no way that a Reactive Variable could have a specific value that is associated with an Apollo Client instance, since it has no way of generally accessing that client.

Now, we do not only have 3 environments - we would also need ways of transporting data from one of those to the next. Since RSC will not necessarily rerender everything from the Layout, we have no way of realiably transporting data from RSC into CC (or at least, you would have to inject extra JSX into every single page).

If we finally had this data flow in place, it would be "RSC -> Browser" or "RSC -> SSR -> Browser only". Strictly unidirectional. So if a variable changed it's value on the Browser, that would never be reflected in your next SSR render - the SSR render would always restart with the initial value of every Reactive Variable.

That last point alone is enough reason for me to say that Reactive Variables on the server make no sense - what benefit do you gain from having variables on the server that almost certainly have a different value than they actually have in the browser later?

Also, I have an extra question here: you always talk about refetching data in reaction to a Reactive Variable change - how would that happen on the server? RSC and SSR never rerender. 
Your components render once and then they are streamed to the browser - any update to a Reactive Variable that would lead to a refetching of data would be too late - the HTML has already been streamed to the browser. And rerendering a React Server Component would be another new request with all variables restarting at their initial value.


All that said, we consider exploring ways of making other things "Reactive". So if some state management library comes up with a good way of keeping Server and Client in sync (maybe with cookies for very small chunks of serializable data?) you could "reactivize" that and plug it into your Apollo Cache instead of using our Browser-Only Reactive Variables.

**Comment by vxcamiloxv** [comment]
I'm no longer using Apollo Client and don't plan to use it in the near future, feel free to open the issue again if you feel it might be relevant, thank you
